### PR TITLE
Add snowflake id

### LINKED ISSUE #3447: Snowflake id
**Is your feature request related to a problem? Please describe.**
Implement snowflake id

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by jackwener**
#### What type of PR is this?
- [ ] bug
- [x] feature

#### Which issue(s) this PR fixes:
close #3447

#### What this PR does / why we need it?

Add a kind of unquie id
  
#### Checklist：
- [x] Documentation affected （If need to modify document, please label it.)
- [ ] Incompatible （If it is incompatile, please describle it and label it.）
- [ ] Need to cherry pick （If need to cherry pick to some branchs, please label the destination version(s).)
- [x] Performance regression: Consumes more CPU
- [x] Performance regression: Consumes more Memory

#### Release notes：
Please confirm whether to reflect in release notes and how to describe:
>                                                                 `

### CODE REVIEW

File: src/common/expression/CMakeLists.txt [hunk 1]
```diff
@@ -32,6 +32,7 @@ nebula_add_library(
     PredicateExpression.cpp
     ListComprehensionExpression.cpp
     ReduceExpression.cpp
+    ../id/SnowFlake.cpp
```
Comment (CPWstatic) [inline src/common/expression/CMakeLists.txt:35]: You could add a new object for the ids.

File: src/common/expression/UUIDExpression.cpp [hunk 2]
```diff
@@ -7,41 +7,28 @@
...
-void UUIDExpression::resetFrom(Decoder& decoder) {
-  // Read field_
-  field_ = decoder.readStr();
-}
+void UUIDExpression::resetFrom(Decoder& decoder) { UNUSED(decoder); }
 
 const Value& UUIDExpression::eval(ExpressionContext& ctx) {
-  // TODO
   UNUSED(ctx);
+  SnowFlake generator;
```
Comment (CPWstatic) [inline src/common/expression/UUIDExpression.cpp:26]: Why always new a generator?

File: src/common/id/SnowFlake.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+      LOG(FATAL) << "Failed to get workerId from meta server";
+    }
+    LOG(INFO) << "Initializing SnowFlake workerId done";
+    workerId_ = result.value().get_workerid();
+  }
+
+  int64_t getId();
+
+ private:
+  mutable std::mutex lock_;
```
Comment (CPWstatic) [inline src/common/id/SnowFlake.h:32]: Why we still have s lock here?

File: src/common/id/SnowFlake.cpp [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  return (timestamp - startStmp_) << timestampLeft | workerId_ << machineLeft | sequence_;
+}
+
+int64_t SnowFlake::getTimestamp() {
+  return std::chrono::duration_cast<std::chrono::milliseconds>(
+             std::chrono::system_clock::now().time_since_epoch())
+      .count();
+}
+
+int64_t SnowFlake::nextTimestamp_() {
```
Comment (CPWstatic) [inline src/common/id/SnowFlake.cpp:40]: We don't use `_`. Please pay attention to the name styles.

File: src/common/id/Snowflake.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      LOG(FATAL) << "Failed to get workerId from meta server";
+    }
+    LOG(INFO) << "Initializing Snowflake workerId done";
+    workerId_ = result.value();
+  }
+
+  static void mockInitWorkerId(int32_t workerId) { workerId_ = workerId; }
+
+  int64_t getId();
+
```
Comment (jackwener) [inline src/common/id/Snowflake.h:32]: should add getInstance() for multiple generators

File: src/common/expression/UUIDExpression.cpp [hunk 6]
```diff
@@ -7,41 +7,28 @@
...
-  // Read field_
-  field_ = decoder.readStr();
-}
+void UUIDExpression::resetFrom(Decoder& decoder) { UNUSED(decoder); }
 
 const Value& UUIDExpression::eval(ExpressionContext& ctx) {
-  // TODO
   UNUSED(ctx);
+  Snowflake generator;
+  result_ = generator.getId();
```
Comment (jackwener) [inline src/common/expression/UUIDExpression.cpp:27]: add a long term generator for client

File: src/meta/processors/id/GetWorkerIdProcessor.cpp [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+ *
+ * This source code is licensed under Apache 2.0 License,
+ */
+
+#include "meta/processors/id/GetWorkerIdProcessor.h"
+
+namespace nebula {
+namespace meta {
+
+void GetWorkerIdProcessor::process(const cpp2::GetWorkerIdReq& req) {
```
Comment (jackwener) [inline src/meta/processors/id/GetWorkerIdProcessor.cpp:11]: add prefix key for multiple generator

File: src/interface/meta.thrift [hunk 8]
```diff
@@ -481,6 +481,17 @@ struct GetPartsAllocResp {
...
 
+// get workerid for snowflake
+struct GetWorkerIdReq {
+    1: binary host, 
+}
+
+struct GetWorkerIdResp {
+    1: common.ErrorCode code,
+    2: common.HostAddr  leader,
+    3: i64              workerid,
```
Comment (jackwener) [inline src/interface/meta.thrift:492]: add prefix key

File: src/daemons/GraphDaemon.cpp [hunk 9]
```diff
@@ -154,6 +154,7 @@ int main(int argc, char *argv[]) {
   gServer->setIOThreadPool(ioThreadPool);
 
   auto interface = std::make_shared<GraphService>();
+  LOG(INFO) << "WJ: Graph Service listening on " << localhost.toString();
```
Comment (jackwener) [inline src/daemons/GraphDaemon.cpp:157]: remove log

File: src/common/id/test/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+
+nebula_add_test(
+  NAME snowflake_test
+  SOURCES SnowflakeTest.cpp
+  OBJECTS
+    $<TARGET_OBJECTS:base_obj>
+    $<TARGET_OBJECTS:snowflake_obj>
+  LIBRARIES gtest gtest_main
+)
```
Comment (jackwener) [inline src/common/id/test/CMakeLists.txt:22]: no new line

File: src/common/id/test/SnowflakeTest.cpp [hunk 11]
```diff
@@ -0,0 +1,28 @@
+/* Copyright (c) 2021 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ */
+
+#include <gflags/gflags.h>
+#include <gtest/gtest.h>
+
+#include <cstdint>
```
Comment (jackwener) [inline src/common/id/test/SnowflakeTest.cpp:9]: redundant

File: src/common/id/Snowflake.cpp [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+  int64_t timestamp = getTimestamp();
+  if (timestamp < lastTimestamp_) {
+    // TODO
+    LOG(FATAL) << "Clock back";
+    return sequence_.load();
+  }
+
+  // if it is the same time, then the microsecond sequence
+  if (lastTimestamp_ == timestamp) {
+    sequence_.store((sequence_.load() + 1) & maxSequenceId_);
```
Comment (yixinglu) [inline src/common/id/Snowflake.cpp:19]: fetch_add

File: src/parser/parser.yy [hunk 13]
```diff
@@ -1091,9 +1095,10 @@ function_call_expression
     ;
 
 uuid_expression
-    : KW_UUID L_PAREN STRING R_PAREN {
-        $$ = UUIDExpression::make(qctx->objPool(), *$3);
-        delete $3;
+    : KW_UUID L_PAREN R_PAREN {
+        // snowflake id init worker in current machine
+        Snowflake::initWorkerId(qctx->getMetaClient());
```
Comment (yixinglu) [inline src/parser/parser.yy:1100]: Should not expose the uuid expression to user

File: src/graph/util/test/CMakeLists.txt [hunk 14]
```diff
@@ -8,6 +8,7 @@ nebula_add_test(
         $<TARGET_OBJECTS:concurrent_obj>
         $<TARGET_OBJECTS:datatypes_obj>
         $<TARGET_OBJECTS:expression_obj>
+    $<TARGET_OBJECTS:snowflake_obj>
```
Comment (yixinglu) [inline src/graph/util/test/CMakeLists.txt:11]: format

File: src/common/id/Snowflake.h [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+ private:
+  /*
+   *  Snowflake id: | timestampBit 41 | workerBit 10 | sequenceBit 12 |
```
Comment (yixinglu) [inline src/common/id/Snowflake.h:32]: It's not enough for our vertex insertion, qps 40w per second now.

File: src/common/expression/test/CMakeLists.txt [hunk 16]
```diff
@@ -144,4 +162,446 @@ nebula_add_executable(
         follybenchmark
         boost_regex
         ${THRIFT_LIBRARIES}
+        ${PROXYGEN_LIBRARIES}
+)
+
+
+nebula_add_test(
+    NAME expression_encode_decode_test
```
Comment (yixinglu) [inline src/common/expression/test/CMakeLists.txt:170]: could cleanup these exprs tests

File: src/common/id/Snowflake.h [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+   */
+  int64_t lastTimestamp_ = -1;          // 41 bits
```
Comment (CPWstatic) [inline src/common/id/Snowflake.h:34]: Init the members like this.
```
int64_t lastTimestamp_{-1};
```

File: src/common/id/Snowflake.h [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+  static void initWorkerId(meta::MetaClient* client) {
+    const std::string& ip = client->getLocalIp();
+    auto result = client->getWorkerId(ip).get();
+    if (!result.ok()) {
+      LOG(FATAL) << "Failed to get workerId from meta server";
+    }
+    workerId_ = result.value();
+  }
+
+  static void mockInitWorkerId(int32_t workerId) { workerId_ = workerId; }
```
Comment (CPWstatic) [inline src/common/id/Snowflake.h:26]: We don't need a interface like this i think, you could use `FREND_TEST` or `friend class` to make your test acces your private members.

File: src/interface/common.thrift [hunk 19]
```diff
@@ -469,6 +469,8 @@ enum ErrorCode {
     E_WRITE_WRITE_CONFLICT            = -3073,
 
     E_CLIENT_SERVER_INCOMPATIBLE      = -3061,
+    // get worker id
+    E_WORKER_ID_FAILED            = -3062,
```
Comment (jievince) [inline src/interface/common.thrift:474]: Format it.

File: src/meta/processors/id/GetWorkerIdProcessor.cpp [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+  const string& ipAddr = req.get_host();
+  auto result = doGet(ipAddr);
+  if (nebula::ok(result)) {
+    string workerIdStr = std::move(nebula::value(result));
+    int32_t workerIdInt32 = std::stoi(workerIdStr);
```
Comment (jievince) [inline src/meta/processors/id/GetWorkerIdProcessor.cpp:16]: The name seems strange.

File: src/meta/processors/id/GetWorkerIdProcessor.h [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+ private:
+  explicit GetWorkerIdProcessor(kvstore::KVStore* kvstore)
+      : BaseProcessor<cpp2::GetWorkerIdResp>(kvstore) {
+    static bool once = [this]() {
+      std::vector<kvstore::KV> data = {{idKey, "0"}};
+      doPut(data);
+      return true;
+    }();
+    UNUSED(once);
+  }
```
Comment (jievince) [inline src/meta/processors/id/GetWorkerIdProcessor.h:31]: Why write in this way?

Comment (jackwener) [reply src/meta/processors/id/GetWorkerIdProcessor.h:31]: init worker id in kvstore

Comment (jievince) [reply src/meta/processors/id/GetWorkerIdProcessor.h:32]: I mean why use lambda?

Comment (jievince) [reply src/meta/processors/id/GetWorkerIdProcessor.h:32]: Ok, got it. Just call it once.

File: src/common/id/Snowflake.h [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+  static constexpr int64_t workerBit = 13;             // worker id bit
+  static constexpr int64_t sequenceBit = 12;           // sequence bit
+
+  static constexpr int64_t maxWorkerId = (1 << workerBit) - 1;  // as worker id mask
+  static constexpr int64_t maxSequence = (1 << sequenceBit) - 1;
+
+  static constexpr int64_t workerIdLeft = sequenceBit;  // workerId left bits
+  static constexpr int64_t timestampLeft = sequenceBit + workerBit;
+
+  static constexpr int64_t firstBitRevert = 0x9000000000000000;  // revert the first bit
```
Comment (jievince) [inline src/common/id/Snowflake.h:57]: See our static variable naming style.

File: src/common/id/Snowflake.cpp [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+// get snowflake id by timestamp
+// update lastTimestamp_ or sequence_
+int64_t Snowflake::getIdByTs(int64_t timestamp) {
+  // if it is the same time, then the microsecond sequence
+  if (lastTimestamp_ == timestamp) {
+    sequence_ = (sequence_ + 1) & maxSequence;
+    // if the microsecond sequence overflow
+    if (sequence_ == 0) {
+      // block to the next millisecond, get the new timestamp
+      timestamp = nextTimestamp();
```
Comment (jievince) [inline src/common/id/Snowflake.cpp:40]: lastTimestamp_ = timestamp;

File: src/common/id/Snowflake.cpp [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+    }
+  } else {
+    lastTimestamp_ = timestamp;
+    sequence_ = 0;
+  }
+  return (timestamp - startStmp) << timestampLeft | workerId_ << workerIdLeft | sequence_;
```
Comment (jievince) [inline src/common/id/Snowflake.cpp:46]: Just put all `lastTimestamp_ = timestamp` here is better.

File: src/interface/meta.thrift [hunk 25]
```diff
@@ -506,6 +506,17 @@ struct GetPartsAllocResp {
...
```
Comment (jievince) [inline src/interface/meta.thrift:517]: i32?

File: src/meta/processors/id/GetWorkerIdProcessor.h [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+
+  void doPut(std::vector<kvstore::KV> data);
+
+  inline static const string idKey = "snowflake_worker_id";
```
Comment (jievince) [inline src/meta/processors/id/GetWorkerIdProcessor.h:36]: static var naming

**Review by CPWstatic — APPROVED** [review]
LGTM.

**Review by jievince — APPROVED** [review]
Good job!

**Review by yixinglu — APPROVED** [review]

**Pull request merged**